keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+  - 'resource' (GVR)
+  - 'name'
+  - 'namespace'
+  - 'operation'
+  - 'userInfo'
+  - 'dryRun'
+  - 'options'
+- 'config' - configuration data of the policy configuration being validated
+
+See below "Reporting violations to Clients" for more detail about how the
+`spec.validations` field works and how violations are reported.
+
+##### Policy Configuration
+
+`PolicyBinding` resources and parameter CRDs together define how cluster
+administrators configure policies for clusters.
+
+Each `PolicyBinding` contains:
+
+- `spec.policy` - A reference to the policy being configured
+- `spec.match` - Match criteria for which resources the policy should validate
+- `spec.params` - Reference to the custom resource containing the params to use
+  when validating resources 
+- `spec.enforcement` - How validation failures should be enforced
+
+Example:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: PolicyBinding
+metadata:
+  name: "xyzlimit-scale.example.com"
+spec:
+  policy: xyzlimit-scale.example.com
+  params: xyzlimit-scale-settings.example.com
+  match:
+    namespaceSelectors:
+    - key: environment,
+      operator: Exists
+  enforcement: [Deny, Audit]
+```
+
+Each parameter CRD defines the custom resources that are referenced by the
+`spec.params` field of `PolicyBinding` resources. Example:
+
+```yaml
+apiVersion: rules.example.com/v1
+kind: XyzLimit
+metadata:
+  name: "xyzlimit-settings.example.com"
+allowed: ["a", "b", "c"]
+banned: ["x", "y", "z"]
+xyz:
+  fuzzFactor: 0.8
+  reticulate: true
+```
+
+Note that for policies that require no parameterization, only the
+`PolicyBinding` is needed.
+
+See "Alternatives considered" section for rejected alternatives. This design was
+selected because:
+
+- Matching criteria is fully defined and validated in a builtin type.
+- Type checking is straight forward.
+- Policy parameterization is separated from the policy binding, allowing for
+  well abstracted parameterization types to be used by applied in different ways
+  by multiple policies.
+
+##### Match Criteria
+
+During admission, the Kubernetes API server will validate the resource being
+admitted against all policy configuration resources that match the resource.
+
+While webhook match rules give a good sense of what types of capabilities might
+be needed, they serve a slightly different purpose.  Webhook match rules make it
+possible to avoid webhook requests, which incur latency and impact availability,
+for resources that don't need to be evaluated by the webhook. CEL expressions
+have a comparatively small impact of latency and are in-process (and so do not
+have the same impact to availability).
+
+For CEL expressions, the primary benefits of match criteria are:
+
+- Match criteria establishes bounds on what sort of admission requests the CEL
+  expressions must consider. The CEL expressions can be written knowing that the
+  match criteria has filtered out requests that is does not need to consider.
```

Comment:
Perhaps, this is an argument in favor of using `match` to simplify CEL expression authoring? Could you leave `match` empty and replicate the arguments inside the CEL expression to achieve the same result? Would having the ability to declare and use local variables also achieve a similar objective when it comes to simplicity of expressions? You run the risk of having two ways of doing the same thing if `match` becomes sufficiently complex. Or does `match` assist with policy indexing and as a side-effect simplify the CEL expression? I think if `match` exists it needs to be very simple and needs to help achieve something that might not be easy to do with CEL directly.

Reply:
Match will be understandable to existing kubernetes authors / configurers. It's going to be necessary to run the match criteria extremely often and it needs to be done in go code. My comment has been lost, but if necessary we need to be able to compile a decision tree about what policies need to be checked given metadata.

Reply:
We did consider not having any "YAML matching" for this feature and instead pushing all matching into CEL. The main deciders for me were:

- Kubernetes already has resource matching as a well established concept
- Match criteria can be built indexed/accelerated/built into decision trees
- Match criteria can evaluate only to true/false. There is no 'error' case to consider
- Match criteria can be used to guide static typing. If the match is for v1.Deployment, we know ahead of runtime what type the `object` variable is

Reply:
@lavalamp @jpbetz would you mind including these key pieces of information in the text above? These sound like much stronger arguments than being familiar and helpful to CEL since they also affect performance and safety in ways that cannot easily be achieved with pure CEL.

Reply:
+1 this is probably worth updating the text over.